nym-api/src/coconut/state.rs

```diff
@@ -102,6 +113,103 @@ impl State {
+                return Err(CoconutError::IncorrectProposal {
+                    reason: String::from("incorrect blinded serial number in description"),
+                });
+            }
+            if !check_proposal(proposal.msgs) {
+                return Err(CoconutError::IncorrectProposal {
+                    reason: String::from("action is not to spend_credential"),
+                });
+            }
+            let client = self.client.clone();
+            tokio::spawn(async move {
+                let ret = client.vote_proposal(id, true, None).await;
+                //SW NOTE: What to do if this fails
+                if let Err(e) = accepted_vote_err(ret) {
+                    log::debug!("Failed to vote on proposal {} - {:?}", id, e);
+                }
+
+                if let Ok(proposal) = client.get_proposal(id).await {
+                    if proposal.status == Status::Passed {
+                        //SW NOTE: What to do if this fails
+                        if let Err(e) = client.execute_proposal(id).await {
+                            log::debug!("Failed to execute proposal {} - {:?}", id, e);
+                        }
+                    }
+                }
+            });
+        }
+        Ok(())
+    }
+
+    pub(crate) async fn refuse_proposal(&self, proposal_id: Option<u64>) {
+        if let Some(id) = proposal_id {
+            let client = self.client.clone();
+            tokio::spawn(async move {
+                //whatever is in the proposal, we can refuse it anyway
+                if let Err(e) = client.vote_proposal(id, false, None).await {
+                    log::debug!("Failed to refuse proposal {:?} - {:?}", id, e)
+                }
+            });
+        }
+    }
+
+    pub(crate) async fn refuse_proposal_online(&self, proposal_id: Option<u64>) {
+        if let Some(id) = proposal_id {
+            //whatever is in the proposal, we can refuse it anyway
+            if let Err(e) = self.client.vote_proposal(id, false, None).await {
+                log::debug!("Failed to refuse proposal {:?} - {:?}", id, e)
+            }
+        }
+    }
+
+    pub(crate) async fn blacklist(&self, public_key: String) {
+        let client = self.client.clone();
+        tokio::spawn(async move {
+            //SW TODO error handling with one log at the end
+            let response = client.propose_for_blacklist(public_key.clone()).await?;
+            let proposal_id = find_proposal_id(&response.logs)
+                .ok_or(CoconutError::ProposalIdError {
+                    reason: "Proposal_ID not found".to_string(),
+                })?
+                .value
+                .parse::<u64>()
+                .map_err(|_| CoconutError::ProposalIdError {
+                    reason: String::from("proposal id could not be parsed to u64"),
+                })?;
+
+            let proposal = client.get_proposal(proposal_id).await?;
+            if proposal.status == Status::Open {
```

Comment:
I did not consider the `Pending`, but the other state mean that we already voted for blacklisting that guy. And it can't be `Refused` since nobody voting against that one